### ISSUE TITLE
Add batching rule for fill_, zero_

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#47755]
## 🚀 Feature

Support efficient computation of fill_ and zero_ inside of vmap.

## To Reproduce
```
>>> import torch
>>> from torch import vmap
>>> x = torch.randn(64, 3)
>>> result = vmap(lambda x: x.fill_(1))(x)
/scratch/rzou/pt/whiteboard-env/bin/ipython:4: UserWarning: Batching rule not implemented for aten::fill_.Scalar falling back to slow (for loop) implementation (Triggered internally
at  ../aten/src/ATen/BatchedFallback.cpp:68.)
```

After this task is complete, we should not see the user warning anymore. Ditto for zero_.

## Context

[`vmap` is a higher-order vectorization operator](https://pytorch.org/docs/master/generated/torch.vmap.html). It takes in a function `f` and returns a new function that maps the `f` over some dimension of the inputs. The new function has similar semantics to performing a for-loop over the mapped dimension:
```py
inputs = torch.randn(N, 3)
inputs_clone = inputs.clone()
f = lambda x: x.fill_(0)
[f(inputs[i]) for i in range(N)] # zero-fill each slice of inputs
expected = torch.vmap(f)(inputs_clone)    # zero-fill all slices of inputs.
assert torch.allclose(inputs, expected)
```
Instead of performing a for-loop, the user could have vectorized their computation by straight up calling `fill_`: `inputs.fill_(0)`. However, for more complicated functions, like `copy_`, vectorizing is not as obvious. In some cases, like efficient Jacobian/Hessian computation, vectorization isn't possible by the user because they cannot control what happens inside of the autograd engine. One of the targeted use cases for vmap is to speed up batched gradient computation in PyTorch.

How vmap works is that a "batching rule" is defined for each operator. The batching rule defines how to perform the operation on inputs with extra dimensions (that we refer to as the "batch dimensions").

## What you should do

Because fill_ is a simple operation, its "batching rule" should just be to call zero_ on the original tensor:
- `[f(inputs[i]) for i in range(N)]`  and `f(inputs)` produce the same result (for `f = lambda x: x.fill_(0)`. Much of the complexity of the task is navigating the abstractions we use to implement batching rules.

More concretely, the task is to implement a BatchedTensor override for the following operators:
- fill_.Scalar
- fill_.Tensor
- zero_

## Resources 

- [How to write a batching rule](https://github.com/pytorch/pytorch/blob/635aebdfab20d9ace8ecfaf347102f0a49fa3553/aten/src/ATen/BatchingRegistrations.cpp#L8-L49)
- [What is a BatchedTensorImpl](https://github.com/pytorch/pytorch/blob/635aebdfab20d9ace8ecfaf347102f0a49fa3553/aten/src/ATen/BatchedTensorImpl.h#L45-L55)
- [What is](https://github.com/pytorch/pytorch/blob/635aebdfab20d9ace8ecfaf347102f0a49fa3553/aten/src/ATen/VmapTransforms.h#L7-L26) [a vmap transform](https://github.com/pytorch/pytorch/blob/635aebdfab20d9ace8ecfaf347102f0a49fa3553/aten/src/ATen/VmapTransforms.h#L41-L47)
- [Sample batching rule implementation](
https://github.com/pytorch/pytorch/blob/635aebdfab20d9ace8ecfaf347102f0a49fa3553/aten/src/ATen/BatchingRegistrations.cpp#L210-L216)
